### PR TITLE
Win loader rebase

### PR DESCRIPTION
**Pull request by parthsane**
Added a Windows Loader and a script to use the Loader.

Currently sgx-tools crate will not compile because of the aesm-client dependency

### CODE REVIEW

**Comment by jethrogb** [comment]
Did you run `cargo update`? I try to avoid that unless a specific update is needed.

**Comment by parthsane** [comment]
Yes, I did a cargo update. I tried rebaseing and it ended up giving errors. 
I didn't know I should avoid a cargo update. I'll do another rebase, this time to ensure I don't have to go through a lot of merge conflicts in Cargo.lock

**Comment by jethrogb** [comment]
This is what I do when I encounter a merge conflict with `Cargo.lock`:
```
git reset Cargo.lock
git checkout Cargo.lock
cargo build
```
You can abort cargo (Ctrl-C) after it prints the first “Compiling” message.

File: sgxs-loaders/src/win_sgx/mod.rs [hunk 1]
```diff
@@ -0,0 +1,342 @@
...
+
+        let mut flags = WinPageProperties::empty();
+        if eadd
+            .secinfo
+            .flags
+            .intersects(SecinfoFlags::PENDING | SecinfoFlags::MODIFIED | SecinfoFlags::PR)
+        {
+            return Err(Error::Add(LibraryError::InvalidParameter));
+        }
+        if eadd.secinfo.flags.intersects(SecinfoFlags::R | SecinfoFlags::W | SecinfoFlags::X) {
```
Comment (jethrogb) [inline sgxs-loaders/src/win_sgx/mod.rs]: This doesn't do anything. My suggestion was to compute the intersection and then test for equality. Also, I just thought of a way to make sure you have all cases covered:
```rust
match (eadd.secinfo.flags & (SecinfoFlags::R | SecinfoFlags::W | SecinfoFlags::X)) as u8 {
    0b000 => ...,
    0b001 => ...,
    etc.
    0b1000...255 => unreachable!(),
}
```

File: sgxs-loaders/src/win_sgx/mod.rs [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+                curhandle,
+                (mapping.base + eadd.offset) as _,
+
+                data.as_ptr() as  *const c_void,
+                0x1000,
+                flags.bits(),
+                std::ptr::null(),
+                0,
+                pDataLoaded,
+                &mut error
```
Comment (jethrogb) [inline sgxs-loaders/src/win_sgx/mod.rs]: Again docs say error is not used.

File: sgxs-loaders/src/win_sgx/mod.rs [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+
+        if flags.contains(WinPageProperties::PAGE_TCS) {
+            // NOTE: For some reason the windows API needs the Read flag set but then removes it
+            assert_eq!(eadd.secinfo.flags.contains(SecinfoFlags::R | SecinfoFlags::W | SecinfoFlags::X), false);
+            flags.insert(WinPageProperties::PAGE_READONLY);
+        }
+        unsafe {
+            let mut error = 0;
+            let mut dataLoaded : usize = 0;
+            let pDataLoaded: *mut usize = &mut dataLoaded;
```
Comment (jethrogb) [inline sgxs-loaders/src/win_sgx/mod.rs]: Unnecessary binding, just pass `&mut data_loaded` below.

File: sgxs-loaders/src/win_sgx/mod.rs [hunk 4]
```diff
@@ -0,0 +1,342 @@
...
+        }
+
+        if flags.contains(WinPageProperties::PAGE_TCS) {
+            // NOTE: For some reason the windows API needs the Read flag set but then removes it
+            assert_eq!(eadd.secinfo.flags.contains(SecinfoFlags::R | SecinfoFlags::W | SecinfoFlags::X), false);
+            flags.insert(WinPageProperties::PAGE_READONLY);
+        }
+        unsafe {
+            let mut error = 0;
+            let mut dataLoaded : usize = 0;
```
Comment (jethrogb) [inline sgxs-loaders/src/win_sgx/mod.rs]: Alternatively:
```
let mut dataLoaded = 0usize;
```
Either way is fine though.

Comment (jethrogb) [inline sgxs-loaders/src/win_sgx/mod.rs]: Don't ignore lint warnings

File: sgxs-loaders/src/win_sgx/mod.rs [hunk 5]
```diff
@@ -0,0 +1,342 @@
...
+            let pDataLoaded: *mut usize = &mut dataLoaded;
+            let curhandle: HANDLE = GetCurrentProcess();
+            let ret = LoadEnclaveData(
+                curhandle,
+                (mapping.base + eadd.offset) as _,
+
+                data.as_ptr() as  *const c_void,
+                0x1000,
+                flags.bits(),
+                std::ptr::null(),
```
Comment (jethrogb) [inline sgxs-loaders/src/win_sgx/mod.rs]: You already have `ptr` imported, so just do `ptr::null`

File: sgxs-loaders/src/win_sgx/mod.rs [hunk 6]
```diff
@@ -0,0 +1,342 @@
...
+use sgxs_crate::einittoken::EinittokenProvider;
+use sgxs_crate::loader;
+use std::fs::OpenOptions;
+
+use std::{mem, ptr};
+mod defs;
+use self::defs::*;
+use std::sync::Arc;
+use crate::{MappingInfo, Tcs};
+use std::path::Path;
```
Comment (jethrogb) [inline sgxs-loaders/src/win_sgx/mod.rs]: Order imports as follows:
1st block: Anything from standard Rust crates (`std`, `core`, `proc_macro`, etc.)
2nd block: Anything from 3rd-party dependencies
3rd block: Anything from our dependencies
4th block: Anything from this crate

File: sgxs-loaders/src/win_sgx/mod.rs [hunk 7]
```diff
@@ -0,0 +1,342 @@
...
+        einittoken: Option<&Einittoken>,
+    ) -> Result<(), Self::Error> {
+        let mut initInfo : ENCLAVE_INIT_INFO_SGX = ENCLAVE_INIT_INFO_SGX {
+            SigStruct : [0 ; 1808],
+            Reserved1 : [0 ; 240],
+            EInitToken: [0 ; 304],
+            Reserved2 : [0 ; 1744],
+        };
+        initInfo.SigStruct.clone_from_slice(&sigstruct.as_ref());
+        if einittoken.is_some() {
```
Comment (jethrogb) [inline sgxs-loaders/src/win_sgx/mod.rs]: `is_some` followed by `unwrap` is an anti-pattern. Whenever you see this, use pattern matching instead:
```rust
if let Some(einittoken) = einittoken {
    // use einittoken
}
```

File: sgxs-loaders/src/win_sgx/mod.rs [hunk 8]
```diff
@@ -0,0 +1,342 @@
...
+        unsafe {
+            let mut error = 0;
+
+            let curhandle: HANDLE = GetCurrentProcess();
+            if InitializeEnclave(
+                curhandle,
+                mapping.base as _,
+                &initInfo as *const _ as *const c_void,
+                4096,
+                Some(&mut error).unwrap()
```
Comment (jethrogb) [inline sgxs-loaders/src/win_sgx/mod.rs]: What are you trying to do here?

File: sgxs-loaders/src/win_sgx/mod.rs [hunk 9]
```diff
@@ -0,0 +1,342 @@
...
+        }
+        unsafe {
+            let mut error = 0;
+
+            let curhandle: HANDLE = GetCurrentProcess();
+            if InitializeEnclave(
+                curhandle,
+                mapping.base as _,
+                &initInfo as *const _ as *const c_void,
+                4096,
```
Comment (jethrogb) [inline sgxs-loaders/src/win_sgx/mod.rs]: use `mem::size_of`

File: sgxs-loaders/src/win_sgx/mod.rs [hunk 10]
```diff
@@ -0,0 +1,342 @@
...
+            let mut error = 0;
+            let mut dataLoaded : usize = 0;
+            let pDataLoaded: *mut usize = &mut dataLoaded;
+            let curhandle: HANDLE = GetCurrentProcess();
+            let ret = LoadEnclaveData(
+                curhandle,
+                (mapping.base + eadd.offset) as _,
+
+                data.as_ptr() as  *const c_void,
+                0x1000,
```
Comment (jethrogb) [inline sgxs-loaders/src/win_sgx/mod.rs]: `data.len()`

File: sgxs-loaders/src/win_sgx/mod.rs [hunk 11]
```diff
@@ -0,0 +1,342 @@
...
+                flags.bits(),
+                std::ptr::null(),
+                0,
+                pDataLoaded,
+                &mut error
+            ) ;
+            if ret == 0 {
+                return Err(Error::Add(LibraryError::OS(IoError::last_os_error())));
+            }
+            assert_eq!(dataLoaded, 0x1000);
```
Comment (jethrogb) [inline sgxs-loaders/src/win_sgx/mod.rs]: compare to `data.len()`

**Comment by jethrogb** [comment]
I think there might be an issue with your clock or timezone setting:
image

**Comment by parthsane** [comment]
Hmm. Windows doesn't take bios time as utc. That's what's causing the time issue.

File: sgxs-loaders/src/lib.rs [hunk 12]
```diff
@@ -16,10 +16,13 @@ extern crate sgxs as sgxs_crate;
 extern crate failure_derive;
 #[macro_use]
 extern crate bitflags;
+#[cfg(windows)] extern crate winapi;
 
 mod generic;
-pub mod isgx;
-pub mod sgx_enclave_common;
+#[cfg(unix)] pub mod isgx;
+#[cfg(unix)] pub mod sgx_enclave_common;
```
[the remark this answers is not in the document]
Comment (parthsane) [reply sgxs-loaders/src/lib.rs]:
```
error[E0425]: cannot find function `mprotect` in module `libc`
   --> sgxs-loaders\src\sgx_enclave_common\mod.rs:248:19
    |
248 |             libc::mprotect(
    |                   ^^^^^^^^ not found in `libc`
```

There are some memory operations like this one which prevents compilation on windows

[the remark this answers is not in the document]
Comment (jethrogb) [reply sgxs-loaders/src/lib.rs]: Ok, just conditionally compile the mprotect operation only on non-windows.

File: sgxs-loaders/src/win_sgx/mod.rs [hunk 13]
```diff
@@ -0,0 +1,342 @@
...
+            return Err(Error::Create(LibraryError::NotSupported));
+        }
+        let base = unsafe {
+            CreateEnclave(
+                curhandle,
+                ptr::null_mut(),
+                ecreate.size as _,
+                0,
+                EnclaveType::Sgx as u32,
+                &secs as *const _ as *const c_void,
```
[the remark this answers is not in the document]
Comment (parthsane) [reply sgxs-loaders/src/win_sgx/mod.rs]: I tired removing the `as *const _ `, but it doesn't compile without it.

**Comment by jethrogb** [comment]
Is this superseded by #89?

**Comment by parthsane** [comment]
Yes. it is superseeded by #89

**Pull request closed without merging**